circuits/src/zk_circuits/valid_fee_redemption.rs

```diff
@@ -0,0 +1,746 @@
+        ElGamalGadget::verify_decryption_key(&witness.recipient_key, &witness.note.receiver, cs)?;
+
+        // --- State Transition --- //
+        Self::verify_wallet_transition(
+            witness.receive_index,
+            &witness.note,
+            &old_wallet,
+            &new_wallet,
+            cs,
+        )
+    }
+
+    /// Verify the state transition of the wallet after settlement
+    fn verify_wallet_transition(
+        receiver_idx: Variable,
+        note: &NoteVar,
+        old_wallet: &WalletVar<MAX_BALANCES, MAX_ORDERS>,
+        new_wallet: &WalletVar<MAX_BALANCES, MAX_ORDERS>,
+        cs: &mut PlonkCircuit,
+    ) -> Result<(), CircuitError> {
+        // All orders should remain the same
+        EqGadget::constrain_eq(&old_wallet.orders, &new_wallet.orders, cs)?;
+
+        // The match fee and managing cluster key should remain the same
+        EqGadget::constrain_eq(&old_wallet.match_fee, &new_wallet.match_fee, cs)?;
+        EqGadget::constrain_eq(&old_wallet.managing_cluster, &new_wallet.managing_cluster, cs)?;
+
+        // The match key should remain the same, the root key may rotate
+        EqGadget::constrain_eq(&old_wallet.keys.pk_match, &new_wallet.keys.pk_match, cs)?;
+
+        // Verify the balance updates
+        let zero_var = cs.zero();
+        let one_var = cs.one();
+        let mut curr_idx = zero_var;
+        let mut found = cs.false_var();
+
+        for (old_bal, new_bal) in old_wallet.balances.iter().zip(new_wallet.balances.iter()) {
+            // Mask the index
+            let is_receive_idx = EqGadget::eq(&curr_idx, &receiver_idx, cs)?;
+            let is_not_receive_idx = cs.logic_neg(is_receive_idx)?;
+            found = cs.logic_or(found, is_receive_idx)?;
+
+            // If the index is the receive index, the mint of the balance should match the
+            // note, otherwise it should match the pre-update wallet mint
+            let matches_note_mint = EqGadget::eq(&new_bal.mint, &note.mint, cs)?;
+            let is_receive_and_note_mint = cs.logic_and(is_receive_idx, matches_note_mint)?;
+            let old_new_mints_equal = EqGadget::eq(&old_bal.mint, &new_bal.mint, cs)?;
+
+            let valid_new_mint = cs.logic_or(is_receive_and_note_mint, old_new_mints_equal)?;
+            cs.enforce_true(valid_new_mint)?;
+
+            // Verify the old balance was either the same mint as the note, or a completely
+            // zero'd balance
+            let was_zerod = EqVecGadget::eq_zero_vec(
+                &[old_bal.amount, old_bal.relayer_fee_balance, old_bal.protocol_fee_balance],
+                cs,
+            )?;
+            let old_mint_equals_note_mint = EqGadget::eq(&old_bal.mint, &note.mint, cs)?;
+            let valid_mint_replacement = cs.logic_or(was_zerod, old_mint_equals_note_mint)?;
+
+            let valid_old_mint = cs.logic_or(is_not_receive_idx, valid_mint_replacement)?;
+            cs.enforce_true(valid_old_mint)?;
```

Comment:
I don't believe these constraints are sufficient to ensure the right mint is being used across the old balance, new balance, and note. Take, for example:

`is_receive_idx = true`
`old_bal.mint = ETH`, o/w zeroed
`new_bal.mint = ETH`
`note.mint = SHITCOIN`

we will have `valid_new_mint == true && valid_old_mint == true`